systemtest/src/test/java/io/strimzi/systemtest/upgrade/kraft/KRaftKafkaUpgradeDowngradeST.java

```diff
@@ -0,0 +1,328 @@
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static io.strimzi.systemtest.TestConstants.KRAFT_UPGRADE;
+import static io.strimzi.test.k8s.KubeClusterResource.kubeClient;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+/**
+ * This test class contains tests for Kafka upgrade/downgrade from version X to X +/- 1, running in KRaft mode.
+ * Metadata for upgrade/downgrade procedure are loaded from kafka-versions.yaml in root dir of this repository.
+ */
+@Tag(KRAFT_UPGRADE)
+public class KRaftKafkaUpgradeDowngradeST extends AbstractKRaftUpgradeST {
+    private static final Logger LOGGER = LogManager.getLogger(KRaftKafkaUpgradeDowngradeST.class);
+
+    private final String continuousTopicName = "continuous-topic";
+    private final int continuousClientsMessageCount = 1000;
+
+    @IsolatedTest
+    void testKafkaClusterUpgrade(ExtensionContext testContext) {
+        List<TestKafkaVersion> sortedVersions = TestKafkaVersion.getSupportedKafkaVersions();
+
+        String producerName = clusterName + "-producer";
+        String consumerName = clusterName + "-consumer";
+
+        for (int x = 0; x < sortedVersions.size() - 1; x++) {
+            TestKafkaVersion initialVersion = sortedVersions.get(x);
+            TestKafkaVersion newVersion = sortedVersions.get(x + 1);
+
+            // If it is an upgrade test we keep the message format as the lower version number
+            String metadataVersion = initialVersion.metadataVersion();
+
+            runVersionChange(initialVersion, newVersion, producerName, consumerName, metadataVersion, 3, 3, testContext);
+        }
+
+        // ##############################
+        // Validate that continuous clients finished successfully
+        // ##############################
+        ClientUtils.waitForClientsSuccess(producerName, consumerName, TestConstants.CO_NAMESPACE, continuousClientsMessageCount);
+        // ##############################
+    }
+
+    @IsolatedTest
+    void testKafkaClusterDowngrade(ExtensionContext testContext) {
+        final TestStorage testStorage = storageMap.get(testContext);
+        List<TestKafkaVersion> sortedVersions = TestKafkaVersion.getSupportedKafkaVersions();
+
+        String clusterName = testStorage.getClusterName();
+        String producerName = clusterName + "-producer";
+        String consumerName = clusterName + "-consumer";
+
+        for (int x = sortedVersions.size() - 1; x > 0; x--) {
+            TestKafkaVersion initialVersion = sortedVersions.get(x);
+            TestKafkaVersion newVersion = sortedVersions.get(x - 1);
+
+            // If it is a downgrade then we make sure to use the lower version number for the message format
+            String metadataVersion = newVersion.metadataVersion();
```

Comment:
I suspect this is wrong? If you have multiple patch releases as today and loop from newest to oldest, you will:
* Start by deploying 3.6.1 and downgrading to 3.6.0
* 3.6.0 will give you metadata version 3.6
* In next iteration, you try to go from 3.6.0 to 3.5.1 and you will fail to downgrade the metadata?

I guess the matadata version here should be always from `sortedVersions.get(0).metadataVersion()`?

Reply:
Yes, correct. It has to be 3.5 to be able to downgrade down to 3.5.1

Reply:
Cool, thanks :)